piptools/utils.py

```diff
@@ -522,3 +531,126 @@ def parse_requirements_from_wheel_metadata(
             markers=parts.markers,
             extras=parts.extras,
         )
+
+
+def callback_config_file_defaults(
```

Comment:
After looking into what this function doers, it became clear why it's so hard to call it properly — it does too much and has side effects.

This should probably be a simple input-to-normalized path pure function.